src/client/proxy/http.cc

```diff
@@ -390,41 +423,98 @@ bool HTTPMessage::ExtractIncomingRequest() {
-  // response?; research this
+bool HTTPMessage::HandleJumpService()
+{
   // TODO(anonimal): add support for remaining services /
   // rewrite this function
-  std::size_t pos1 = m_Path.rfind(m_JumpService.at(0));
-  std::size_t pos2 = m_Path.rfind(m_JumpService.at(1));
-  std::size_t pos;
-  if (pos1 == std::string::npos) {
-    if (pos2 == std::string::npos)
-      return;  // Not a jump service
-    else
-      pos = pos2;
-  } else {
-    if (pos2 == std::string::npos)
-      pos = pos1;
-    else if (pos1 > pos2)
-      pos = pos1;
-    else
-      pos = pos2;
-  }
-  auto base64 = m_Path.substr(pos + m_JumpService.at(0).size());
-  // We must decode
-  base64 = boost::network::uri::decoded(base64);
-  // Insert into address book
-  LOG(debug)
-    << "HTTPProxyHandler: jump service for " << m_Address
-    << " found at " << base64 << ", inserting to address book";
-  // TODO(unassigned): this is very dangerous and broken.
-  // We should ask the user for confirmation before proceeding.
-  // Previous reference: http://pastethis.i2p/raw/pn5fL4YNJL7OSWj3Sc6N/
-  // We *could* redirect the user again to avoid dirtiness in the browser
-  kovri::client::context.GetAddressBook().InsertAddressIntoStorage(
-      m_Address, base64);
-  m_Path.erase(pos);
+
+  // Perform sanity check again to:
+  // - ensure valid jump service request
+  // - update pos with m_JumpService sub-string index (hacky)
+  std::size_t pos = 0;
+  if (!IsJumpServiceRequest(pos))
+    {
+      LOG(error) << "HTTPProxyHandler: not a valid jump service request";
+      // Jump service parameter not found in URL, so just return
+      return false;
+    }
+
+  if (!ExtractBase64Destination(pos))
+    {
+      LOG(error) << "HTTPProxyHandler: unable to process base64 destination for "
+                 << m_Address;
+      m_URL.erase(pos);
+      return false;
+    }
+
+  LOG(debug) << "HTTPProxyHandler: jump service for " << m_Address
+             << " found at " << m_Base64Destination;
+  m_URL.erase(pos);
+  return true;
+}
+
+// Check if request is for a jump service
+bool HTTPMessage::IsJumpServiceRequest(std::size_t& pos)
+{
+  std::size_t pos1 = m_URL.rfind(m_JumpService.at(0));
+  std::size_t pos2 = m_URL.rfind(m_JumpService.at(1));
+  if (pos1 == std::string::npos)
+    {
+      if (pos2 == std::string::npos)
+        return false;  // Not a jump service
+      else
+        pos = pos2;
+    }
+  else
+    {
+      if (pos2 == std::string::npos)
+        pos = pos1;
+      else if (pos1 > pos2)
+        pos = pos1;
+      else
+        pos = pos2;
+    }
+
+  // final sanity check
+  // TODO(oneiric): may be unnecessary, consider removing
+  if (pos != std::string::npos && pos != 0)
+    return true;
+  else
+    return false;
+}
+
+// Extract & uri-decode base64-encoded I2P destination
+bool HTTPMessage::ExtractBase64Destination(std::size_t pos)
+{
+  std::size_t base64_size = pos + m_JumpService.at(0).size();
+  if (pos > 0 && base64_size < m_URL.size())
```

Comment:
`pos` is unsigned and will never be negative. Simply test if not null `if (pos && base64_size < M_URL.size())`